### ISSUE TITLE
Library bundling at build time.

### ISSUE THREAD

**Issue by ejgallego** [coq/coq#6208]
The set of libraries bundled with Coq should not be the same set that is hosted in the main github repos.

For example, bignums is not living the Coq repository, but it makes sense to build binaries that include it.

Thus, we should have a build-system flag that will download and bundle specific versions of packages.

**Comment by Zimmi48** [comment]
OPAM users can easily download and compile plugins such as bignums and that is also true for Nix users but this is not the case for pre-built package users (in particular for Windows users), so it does make sense indeed to ship more in pre-built packages.
Another solution would be to automatically build and distribute binary packages for plugins in the Coq package index. They would require specific installation steps but would be as easy to install as the main package. I don't know how realistic this is.

**Comment by maximedenes** [comment]
I support such a model. I would even call it a Coq distribution.

> it makes sense to build binaries that include it.

Should we also distribute a source distribution including those plugins?

**Comment by mattam82** [comment]
I agree we should have a CoqLive kind of distribution as well, especially for windows and OS X users. I’m wondering what @soegstrup thinks about the feasibility of this. Another “direction” was the proposed use of images that Yves pursued a little bit with Docker, but I guess that’s not incompatible.

**Comment by mattam82** [comment]
I fear trying to solve the fine-grain problem of individual multi-platform versioned binary packages is going to be hell and that’s what packages managers are for.

**Comment by Zimmi48** [comment]
> Should we also distribute a source distribution including those plugins?

What would be the point of distributing a source distribution? (Apart possible legal reasons.) Is anyone going to use it?

**Comment by maximedenes** [comment]
Apart from legal reasons, I don't know, but I was asking in case someone had a use case in mind.

**Comment by Zimmi48** [comment]
Let's try if we can manage to have this for 8.7.1.

**Comment by JasonGross** [comment]
> What would be the point of distributing a source distribution? 

I seem to recall company-coq having a "jump to definition" feature (though I may be wrong) that only works if the .v file is around.

**Comment by Zimmi48** [comment]
Did someone inform @MSoegtropIMC of this plan?

It is unlikely though to have this ready for 8.7.1 (which is due on Friday).

**Comment by MSoegtropIMC** [comment]
I must admit I didn't fully understand what the requirements / goals are.

Is it something like:

- build several additional libraries when building a Windows delivery (using my scripts)
- only include parts of these in the installer and have others automatically downloadable from a server
- make it selectable if the installer installs / downloads sources or not
- possibly being able to choose different versions
- make the installer such that packages can be easily added / removed after the initial install

All this wouldn't be complicated, but I wouldn't do it with the NSIS based installer. If you want to do anything more advanced, I would switch to WIX (Windows Installer XML toolkit). I might be able to do this over the weekend for some packages, but it is fairly short notice.

Best regards,

Michael

**Comment by Zimmi48** [comment]
Dear Michael,

Very sorry for the short notice. I was counting on @mattam82 and @maximedenes to get in touch with you about this question; I should have done it myself from the beginning.

I think everything you propose would be wonderful, but a subset would also be acceptable as well if it is less work (in fact the first point is the most important to start with, the rest could come later on). No opposition to switching to a more advantaged installer toolkit on my side. You're the expert ;-)

If it allows you to do it less in a rush, we could aim at having this for 8.7.2 (February, 15th), or 8.8+beta1 (March, 15th).

**Comment by maximedenes** [comment]
I like the idea, but let's be clear: this is not going to happen for 8.7.1. We are already under pressure to solve blockers in time, so let's take the time to do this properly. So @MSoegtropIMC, I suggest we synchronize at some point (probably in January if that's ok for you), but there is certainly no rush.

**Comment by MSoegtropIMC** [comment]
Then I would suggest to bundle this with the MinGW OCaml, buildtools and make I am working on and plan to release January. The main purpose of bundling OCaml and buildtools is to allow native compute, but this should also allow to compile libraries with OCaml modules with an installed Coq system (without Cygwin). I am not sure I can achieve this in the first step in January, but I think this topic has a lot of overlap with ease of library usage and we should work on a story taking this into account.

Btw.: the main ingredient for the above is a mini Posix shell written in C and replacing the OCaml system command and used by make.

**Comment by Zimmi48** [comment]
Hi @MSoegtropIMC,
Did you make progress on your planned mini-shell + ocaml build tools?
Best regards,
Théo

**Comment by Zimmi48** [comment]
Even if the new OCaml build tools are not ready, it would be good if we could still distribute a bundled package of Coq and some external plugins / libraries for 8.8. As a first step, it doesn't need to be very flexible. This could just be an installer that contains all the plugins / libraries we chose. @gares may have some expertise on this topic as well since he has done that for math-comp I believe.

**Comment by MSoegtropIMC** [comment]
Dear Théo

if you specify a list of libraries, I am happy to add this to my scripts quickly, including some easy to extend mechanisms in my scripts.
I would suggest to make such libraries optional modules in the installer (that is the user can chose which to install).
The only issue I see is the build time in the regression tests - there is an upper limit and chaining jobs seems to be tricky because there is no proper artifact server. But I didn't follow progress in this area in the last 3 months.

Best regards,

Michael

**Comment by MSoegtropIMC** [comment]
P.S.: another option would be to supply a script which sets up a cygwin environment which can be used to build libraries in conjunction with an installed Coq system.

Best regards,

Michael

**Comment by Zimmi48** [comment]
> The only issue I see is the build time in the regression tests - there is an upper limit and chaining jobs seems to be tricky because there is no proper artifact server. But I didn't follow progress in this area in the last 3 months.

Indeed. Right now, the build time for the packaging jobs is 30-40 minutes, so adding bignums (2 min build), equations (4 min build) or ltac2 (1 min build) shouldn't be a problem but adding larger libraries could be problematic. AppVeyor has artifact storage (and we use it to get the installers) but I don't know whether it has artifact sharing between jobs. Anyways, I think it would be great if we could start packaging the three plugins I just mentioned (libraries are less a problem on Windows). In the future, we could have other stuff, like the parametricity plugin, elpi and template-coq but right now they are not tracked in our CI so this is not an option.

**Comment by gares** [comment]
> This could just be an installer that contains all the plugins / libraries we chose. @gares may have some expertise on this topic as well since he has done that for math-comp I believe.

Making an installer for windows for "user-contrib" stuff is very easy. In math-comp I simply build and install and then the run the script https://github.com/math-comp/math-comp/blob/master/etc/win-installer.nsi (it requires NSIS to be installed somewhere and the precise invocation is in the comments).  Such a script needs to be fixed adding back the lines about `*.cm*` files that I've removed when ssr was merged in Coq).

**Comment by gares** [comment]
In other words, if you give me a `...coq\lib\user-contrib\` folder with all the stuff you need, I'm happy to write a `.nsi` file that bundles that.

**Comment by MSoegtropIMC** [comment]
Would it be better to have separate installers, or should at least elementary components be bundled in the main installer?

**Comment by Zimmi48** [comment]
Both roads have their advantages: a single installer means you have to download it only once and you only get "approved" plugins while multiple installers mean we could distribute some stuff for plugin developers to easily set up AppVeyor and automatically build their own installers.

**Comment by gares** [comment]
My feeling is that it would be more practical for the user to just download 1 installer.

*But* most of the packages mentioned above are moving targets. I've one of my own (elpi) that could enter the picture. As the upstream of elpi I'd be happy to have it easy-to-install by users, but I'd be also interested in making my own releases (imagine I find a very nasty bug).

IMHO additions to the Coq installer should be carefully discussed with the author of the extension, and in any case, only very stable extensions should be shipped.

**Comment by gares** [comment]
> we could distribute some stuff for plugin developers to easily set up AppVeyor and automatically build their own installers.

+1

**Comment by gares** [comment]
PR #6881 provides some support for this on windows

**Comment by maximedenes** [comment]
@ejgallego I think setting the milestones should be the responsibility of the assignee. You can't write their schedule.

**Comment by gares** [comment]
Why is this assigned to me, by the way?

**Comment by Zimmi48** [comment]
Because you fixed it, at least partially for bignums. However, I guess this should be assigned back to @MSoegtropIMC since he has bigger plans for this.

**Comment by MSoegtropIMC** [comment]
Please feel free to assign it to me. I anyway wanted to work on the signature verification next week.

**Comment by ejgallego** [comment]
IMHO we can close this; we have some proposals already working [such as the windows install] and for Linux/OSX Dune / OPAM  should do the rest.

**Comment by MSoegtropIMC** [comment]
At least from Windows side I would consider this to be fixed. Since @ejgallego says that it is fixed with Dune on Linux/OSX as well, I will close this.

**Comment by Zimmi48** [comment]
I don't get how dune / opam fixes this on Linux / macOS. My impression is that opam is too hard to use for casual Coq users, and we should provide a distribution of Coq with packages on Linux and macOS too, or a nicer wrapper around opam. But anyways, I see this as a long term plan and don't mind this issue being closed.

**Comment by ejgallego** [comment]
With Dune or Opam you can compose package and have a uniform build just by providing a list of repos; IMHO the issue for OSX / Linux is how to provide binary packages, which IMO is a separate issue.

**Comment by MSoegtropIMC** [comment]
I don't think it makes sense to provide binary Linux packages. In my experience this would have to be done separately for the major distributions (at least 3 or 4, maybe more). Otherwise the maintenance effort from users trying to install something and failing can be very substantial. I am privately using debian 9 and use some commercial software which is distributed as binary and must say that it frequently requires quite a bit of expertise and manual tweaking to get these installed at all, even though the providers do have separate binaries for Debian, Redhat and Suse.

I would think on Linux good noob approved instructions would make more sense.

(My son volunteers to do the Linux noob test)

**Comment by Zimmi48** [comment]
@MSoegtropIMC Let's have this discussion at #7161 and #9289 instead.
